### PR TITLE
8236671: NullPointerException in JKS keystore

### PR DESCRIPTION
**Pull request by coffeys**
Trivial enough change. Improved the exception thrown from JceKeyStore also.

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Issue
 * [JDK-8236671](https://bugs.openjdk.java.net/browse/JDK-8236671): NullPointerException in JKS keystore

### Reviewers
 * [Hai-May Chao](https://openjdk.java.net/census#hchao) (@haimaychao - Committer) ⚠️ Review applies to 836ea7e7cfc187f69886f184a90f101c82c256b3
 * [Xue-Lei Andrew Fan](https://openjdk.java.net/census#xuelei) (@XueleiFan - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.java.net/jdk pull/3588/head:pull/3588` \
`$ git checkout pull/3588`

Update a local copy of the PR: \
`$ git checkout pull/3588` \
`$ git pull https://git.openjdk.java.net/jdk pull/3588/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 3588`

View PR using the GUI difftool: \
`$ git pr show -t 3588`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.java.net/jdk/pull/3588.diff">https://git.openjdk.java.net/jdk/pull/3588.diff</a>

</details>

### CODE REVIEW

**Review by haimaychao — APPROVED** [review]

**Comment by XueleiFan** [comment]
It looks like a public behavior change to me.  Did you want to file a CSR and update the specification (KeyStore) as well?

**Comment by coffeys** [comment]
> It looks like a public behavior change to me. Did you want to file a CSR and update the specification (KeyStore) as well? I think it would be nice if we could keep use the old exception, IllegalArgumentException, as described in the bug.

@XueleiFan  - The spec in question has been broken for almost 3 years with the throwing of NPE.

One issue here is that Sun provider with JKS keystore will throw IllegalArgumentException in older JDK versions but the SunJCE provider and JCEKS keystore throws KeyStoreException when null password is encountered . There's a mismatch. To me, it looks like KeyStoreException is the correct exception in such scenarios (and according to API spec)

I can file a CSR to have the implementation adhere to spec if that's desired.

**Comment by XueleiFan** [comment]
> > It looks like a public behavior change to me. Did you want to file a CSR and update the specification (KeyStore) as well? I think it would be nice if we could keep use the old exception, IllegalArgumentException, as described in the bug.
> 
> @XueleiFan - The spec in question has been broken for almost 3 years with the throwing of NPE.
> 
> One issue here is that Sun provider with JKS keystore will throw IllegalArgumentException in older JDK versions but the SunJCE provider and JCEKS keystore throws KeyStoreException when null password is encountered . There's a mismatch. To me, it looks like KeyStoreException is the correct exception in such scenarios (and according to API spec)
> 
> I can file a CSR to have the implementation adhere to spec if that's desired.

It makes sense to me.  I think it would be good to have this stated in the spec in case more mismatch introduced in the future.

File: src/java.base/share/classes/sun/security/provider/JavaKeyStore.java [hunk 1]
```diff
@@ -287,6 +287,9 @@ public void engineSetKeyEntry(String alias, Key key, char[] password,
                 entry.date = new Date();
 
                 // Protect the encoding of the key
+                if (password == null) {
```
Comment (seanjmullan) [inline src/java.base/share/classes/sun/security/provider/JavaKeyStore.java:290]: You could move this check earlier, before the try block.

**Comment by coffeys** [comment]
/csr

I'll log a CSR to highlight the behavioural difference/correction being made to the SUN JKS type keystore.

**Review by XueleiFan — APPROVED** [review]

**Comment by coffeys** [comment]
KeyStore specification will be tightened up via another bug record: https://bugs.openjdk.java.net/browse/JDK-8266351

**Pull request closed without merging**